tensorflow_addons/optimizers/discriminative_layer_training_test.py

```diff
@@ -0,0 +1,271 @@
+    other models returned by this function, for the duration of that
+    continuous integration run
+
+    model is intended to work with
+    x = np.ones(shape = (None, 32, 32, 3), dtype = np.float32)
+    y = np.zeros(shape = (None, 5), dtype = np.float32)
+    y[:, 0] = 1.
+    """
+
+    cnn_model_path = "cnn.h5"
+
+    if not os.path.exists(cnn_model_path):
+        # force eager mode for simple initialization of vars
+        with context.eager_mode():
+            tf.random.set_seed(1)
+            bignet = tf.keras.applications.mobilenet_v2.MobileNetV2(
+                include_top=False, weights=None, input_shape=(32, 32, 3), pooling="avg"
+            )
+
+            # take the first few layers so we cover BN, Conv, Pooling ops for testing
+            net = tf.keras.models.Model(
+                inputs=bignet.input, outputs=bignet.get_layer("block_2_add").output
+            )
+            model = tf.keras.Sequential(
+                [
+                    net,
+                    tf.keras.layers.GlobalAveragePooling2D(),
+                    tf.keras.layers.Dropout(0.5),
+                    tf.keras.layers.Dense(5, name="head"),
+                ]
+            )
+            # always save and never return initialized model from memory
+            # it seems you cannot pass variables from a nested eager context to its parent graph context
+
+            model.save(cnn_model_path)
+
+    # load the initialized model from the disk
+    return tf.keras.models.load_model(cnn_model_path)
+
+
+# TODO: get toy_run to work
+def toy_rnn():
+    """
+
+    Consistently create model with same random weights
+    skip head activation to allow both bce with logits and cce with logits
+    intended to work with
+
+    x = np.ones(shape = (None, 32, 32, 3), dtype = np.float32)
+    y = np.zeros(shape = (None, 5), dtype = np.float32)
+    y[:, 0] = 1.
+    """
+
+    tf.random.set_seed(1)
+
+    model = tf.keras.Sequential()
+    model.add(tf.keras.layers.Input(shape=(32, 32, 3)))
+    model.add(tf.keras.layers.Reshape(target_shape=(32, 96)))
+    model.add(tf.keras.layers.Cropping1D(cropping=(0, 24)))
+    model.add(tf.keras.layers.Bidirectional(tf.keras.layers.LSTM(8)))
+    model.add(tf.keras.layers.Dropout(0.5))
+    model.add(tf.keras.layers.Dense(5))
+
+    return model
+
+
+def get_train_results(model, verbose=False):
+    """Run a traininng loop and return the results for analysis
+    model must be compiled first
+    """
+    tf.random.set_seed(1)
+    x = np.ones(shape=(32, 32, 32, 3), dtype=np.float32)
+    y = np.zeros(shape=(32, 5), dtype=np.float32)
+    y[:, 0] = 1.0
+
+    return model.fit(x, y, epochs=10, batch_size=16, verbose=verbose, shuffle=False)
+
+
+def zipped_permutes():
+    model_fns = [toy_cnn]
+    losses = [
+        # tf.keras.losses.BinaryCrossentropy(from_logits=True),
+        tf.keras.losses.CategoricalCrossentropy(from_logits=True),
+        #         tf.keras.losses.MeanSquaredError(),
```

Comment:
I'm not sure iterating over losses should really test anything in this?

Reply:
I think you're right. Losses would determine the gradients that are passed to the optimizers, which would be unaffected by the wrapper. The compute gradient step stays exactly the same as before and only the apply gradients is modified by the wrapper. 

I'll use only categorical going forward